### PR TITLE
[#5224] Allow to use Unsafe.reallocateMemory(...) in UnpooledUnsafeNo…

### PR DESCRIPTION
**Pull request by normanmaurer**
…CleanerDirectByteBuf.

Motivation:

If the user uses unsafe direct buffers with no cleaner we can use Unsafe.reallocateMemory(...) as optimization when we need to expand the buffer.

Modifications:

Use Unsafe.relocateMemory(...) in UnpooledUnsafeNoCleanerDirectByteBuf.

Result:

Less expensive expanding of buffers.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@trustin PTAL... This depends on https://github.com/netty/netty/pull/5314

File: buffer/src/main/java/io/netty/buffer/UnpooledUnsafeNoCleanerDirectByteBuf.java [hunk 1]
```diff
@@ -34,4 +34,38 @@ protected ByteBuffer allocateDirect(int initialCapacity) {
...
+        if (newCapacity > oldCapacity) {
+            ByteBuffer oldBuffer = buffer;
+            ByteBuffer newBuffer = PlatformDependent.reallocateDirectNoCleaner(oldBuffer, newCapacity);
+            setByteBuffer(newBuffer, false);
+        } else if (newCapacity < oldCapacity) {
+            ByteBuffer oldBuffer = buffer;
+            ByteBuffer newBuffer = allocateDirect(newCapacity);
+            if (readerIndex < newCapacity) {
+                if (writerIndex > newCapacity) {
+                    writerIndex(writerIndex = newCapacity);
```
Comment (netkins) [inline buffer/src/main/java/io/netty/buffer/UnpooledUnsafeNoCleanerDirectByteBuf.java]: MAJOR Extract the assignment out of this expression. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AAssignmentInSubExpressionCheck)

**Comment by Scottmitch** [comment]
I'll take a look after #5314 is merged

**Comment by Scottmitch** [comment]
@normanmaurer - what is the status of this PR? was this merged ... delete the branch?

**Comment by normanmaurer** [comment]
@Scottmitch nope will re-open a new one as this was automatically closed when merged the depending one :(

**Comment by normanmaurer** [comment]
See https://github.com/netty/netty/pull/5350

**Pull request closed without merging**